File: frontend/app/_lib/analytics.ts
import { apiFetch } from "./api";

export interface HeatmapDay {
  date: string;
  count: number;
}

export interface DashboardData {
  heatmap: HeatmapDay[];
  current_streak: number;
  longest_streak: number;
  total_solved: number;
  cf_rating: number | null;
  has_verified_handle: boolean;
  is_syncing: boolean;
}

export interface TagStat {
  tag: string;
  solved_count: number;
  attempt_count: number;
  acceptance_rate: number;
  last_activity_at: string | null;
}

export interface RatingEntry {
  cf_contest_id: number;
  contest_name: string;
  old_rating: number;
  new_rating: number;
  delta: number;
  rank: number;
  contest_time: string;
}

export type WeaknessSignalType = "low_success" | "neglected" | "under_practiced";

export interface WeaknessSignal {
  id: string;
  tag: string;
  signal_type: WeaknessSignalType;
  score: number;
  reason: string;
  computed_at: string;
}

export interface Recommendation {
  id: string;
  problem_id: string;
  problem_name: string;
  tag: string;
  difficulty: number;
  url: string;
  reason: string;
  position: number;
}

export interface RecommendationSet {
  id: string;
  generated_at: string;
  recommendations: Recommendation[];
}

export async function fetchDashboard(token: string): Promise<DashboardData> {
  const res = await apiFetch("/api/v1/analytics/dashboard", { token });
  if (!res.ok) throw new Error("fetch dashboard failed");
  return res.json();
}

export async function fetchTags(token: string): Promise<TagStat[]> {
  const res = await apiFetch("/api/v1/analytics/tags", { token });
  if (!res.ok) throw new Error("fetch tags failed");
  return res.json();
}

export async function fetchRatingHistory(token: string): Promise<RatingEntry[]> {
  const res = await apiFetch("/api/v1/analytics/rating-history", { token });
  if (!res.ok) throw new Error("fetch rating history failed");
  return res.json();
}

export async function fetchWeaknesses(token: string): Promise<WeaknessSignal[]> {
  const res = await apiFetch("/api/v1/analytics/weaknesses", { token });
  if (!res.ok) throw new Error("fetch weaknesses failed");
  return res.json();
}

export async function fetchRecommendations(token: string): Promise<RecommendationSet | null> {
  const res = await apiFetch("/api/v1/analytics/recommendations", { token });
  if (!res.ok) throw new Error("fetch recommendations failed");
  return res.json();
}

export async function refreshRecommendations(token: string): Promise<RecommendationSet | null> {
  const res = await apiFetch("/api/v1/analytics/recommendations/refresh", {
    token,
    method: "POST",
  });
  if (!res.ok) throw new Error("refresh recommendations failed");
  return res.json();
}

File: backend/app/schemas/analytics.py
import uuid
from datetime import datetime

from pydantic import BaseModel

from app.models.signals import WeaknessSignalType


class HeatmapDay(BaseModel):
    date: str   # ISO format YYYY-MM-DD
    count: int  # total submissions (any verdict) — drives heatmap intensity, matches CF behavior
    solved: int # accepted submissions — shown in tooltip alongside total


class DashboardResponse(BaseModel):
    heatmap: list[HeatmapDay]
    current_streak: int
    longest_streak: int
    total_solved: int
    cf_rating: int | None
    has_verified_handle: bool
    is_syncing: bool = False  # True when handle sync is in-progress or has never completed


class TagStatsResponse(BaseModel):
    tag: str
    solved_count: int
    attempt_count: int
    acceptance_rate: float
    last_activity_at: datetime | None

    model_config = {"from_attributes": True}


class RatingHistoryResponse(BaseModel):
    cf_contest_id: int
    contest_name: str
    old_rating: int
    new_rating: int
    delta: int
    rank: int
    contest_time: datetime

    model_config = {"from_attributes": True}


class WeaknessSignalResponse(BaseModel):
    id: uuid.UUID
    tag: str
    signal_type: WeaknessSignalType
    score: float
    reason: str
    computed_at: datetime

    model_config = {"from_attributes": True}


class RecommendationResponse(BaseModel):
    id: uuid.UUID
    problem_id: str
    problem_name: str
    tag: str
    difficulty: int
    url: str
    reason: str
    position: int

    model_config = {"from_attributes": True}


class RecommendationSetResponse(BaseModel):
    id: uuid.UUID
    generated_at: datetime
    recommendations: list[RecommendationResponse]

    model_config = {"from_attributes": True}
